data_extractor/google_semantic_location_history/__init__.py

```diff
@@ -0,0 +1,106 @@
+"""Script to extract data from Google Semantic History Location zipfile"""
+__version__ = '0.1.0'
+
+import json
+import collections
+import itertools
+import re
+import zipfile
+
+import pandas as pd
+
+
+def __visit_duration(data):
+    """Get duration per visited place
+    Args:
+        data (dict): Google Semantic Location History data
+    Returns:
+        dict: duration per visited place, sorted in descending order
+    """
+    placevisit_duration = []
+    for data_unit in data["timelineObjects"]:
+        if "placeVisit" in data_unit.keys():
+            address = data_unit["placeVisit"]["location"]["placeId"]
+            start_time = data_unit["placeVisit"]["duration"]["startTimestampMs"]
+            end_time = data_unit["placeVisit"]["duration"]["endTimestampMs"]
+            placevisit_duration.append(
+                {address: (int(end_time) - int(start_time))/(1e3*24*60*60)})
+
+    # list of places visited
+    address_list = {list(duration.keys())[0] for duration in placevisit_duration}
+
+    # dict of time spend per place
+    places = {}
+    for address in address_list:
+        places[address] = round(sum(
+            [duration[address] for duration in placevisit_duration \
+                if address == list(duration.keys())[0]]), 3)
+    # Sort places to amount of time spend
+    places = collections.OrderedDict(
+        sorted(places.items(), key=lambda kv: kv[1], reverse=True))
+
+    return places
+
+def __activity_duration(data):
+    """Get total duration of activities
+    Args:
+        data (dict): Google Semantic Location History data
+    Returns:
+        float: duration of actitvities in days
+    """
+    activity_duration = 0.0
+    for data_unit in data["timelineObjects"]:
+        if "activitySegment" in data_unit.keys():
+            start_time = data_unit["activitySegment"]["duration"]["startTimestampMs"]
+            end_time = data_unit["activitySegment"]["duration"]["endTimestampMs"]
+            activity_duration += (int(end_time) - int(start_time))/(1e3*24*60*60)
+    return activity_duration
+
+def process(file_data):
+    """Return relevant data from zipfile for years and months
+    Args:
+        file_data: zip file or object
+
+    Returns:
+        dict: dict with summary and DataFrame with extracted data
+    """
+    years = [2020, 2021]
+    months = ["JANUARY"]
+    results = []
+    filenames = []
+    name = None
+
+    # Extract info from selected years and months
+    with zipfile.ZipFile(file_data) as zfile:
+        for year in years:
+            for month in months:
+                for name in zfile.namelist():
```

Comment:
This nested construct calls `zfile.namelist()` many times. An optimization would be to get the names and then check if they match the `year_month.json` pattern. This could be done with a regex like: `match = re.match(r"(\d{4})_(\w+).json", name)`. This can then be split into year and month with: `year, month = match.groups()`. The year has to be converted to an int after that.

Reply:
I couldn't get it to work, so compromised by getting the list outside of the loop. You can also make code change suggestion if you checked that it worked.